In [139]:
import pandas as pd
import numpy as np
import requests

import matplotlib.pyplot as plt
import seaborn as sns

## Exercise 1

**Write a Python function that can be used to query data from the Worldbank Indicator API. Your function should:**

• take the following input parameters: indicators, countries, and years.

• return a Pandas DataFrame of the queried data

• have a docstring that explains what the function does, what the input parameters are, and what
the output is

• minimize the number of API calls necessary to retrieve the data
Demonstrate that your function works by querying the following data (codes are provided in parentheses):

a) The total population (SP.POP.TOTL) of Germany (DE) and France (FR) between 2015 and 2020.

b) The total population (SP.POP.TOTL), GDP in current US$ (NY.GDP.MKTP.CD), and life ex-
pectancy in years at birth (SP.DYN.LE00.IN) of all countries (all) in 2012. Print the shape of the
resulting DataFrame and display its first 10 rows.

c) State how many API calls your function makes for a) and b) respectively.
Notes:

• To solve the exercises study the documentation of the basic call structures. Most of the information
you need is provided there.

• If needed, additional information about the API is available here. For instance, you will find links
to the list of available indicators and countries, and explanations on error codes.

• Note that by default the API r


In [140]:
call_count = 0 
def call_api(indicators,countries,years,) -> pd.DataFrame:
    """
    Call the World Bank API to get data for the specified indicators, countries, and years.
    
    Parameters:
    - indicators: List of indicator codes
    - countries: List of country codes
    - years: List of years
    
    Returns:
    - DataFrame with the requested data
    """
    global call_count
    if isinstance(indicators, list):
        indicators = ';'.join(indicators)
    if isinstance(countries, list):
        countries = ';'.join(countries)
    
    
    url = f"http://api.worldbank.org/v2/country/{countries}/indicator/{indicators}?date={years}&format=json&per_page=10000&source=2"
    print("url: ",url)

      # Make the API request
    response = requests.get(url)
    
    # Check for HTTP errors
    if response.status_code != 200:
        print(f"Error: Received status code {response.status_code}")
        print(f"Response text: {response.text}")
        return None
    call_count += 1

    try:
        data = response.json()
    except ValueError as e:
        print("Error decoding JSON:", e)
        print("Response text:", response.text)
        return None
    
    # Check if the response contains data
    if len(data) < 2 or data[1] is None:
        print("No data available in the response.")
        print("Response content:", data)  # Debugging: Print the full response
        return None
    
    # Convert the JSON data to a DataFrame
    df = pd.json_normalize(data[1])
    return df
    
df= call_api(
    indicators="SP.POP.TOTL",
    countries=["DE","FR"],
    years="2015:2020"
)

df


url:  http://api.worldbank.org/v2/country/DE;FR/indicator/SP.POP.TOTL?date=2015:2020&format=json&per_page=10000&source=2


countryiso3code  date     value unit obs_status  decimal indicator.id  \
0              DEU  2020  83160871                        0  SP.POP.TOTL   
1              DEU  2019  83092962                        0  SP.POP.TOTL   
2              DEU  2018  82905782                        0  SP.POP.TOTL   
3              DEU  2017  82657002                        0  SP.POP.TOTL   
4              DEU  2016  82348669                        0  SP.POP.TOTL   
5              DEU  2015  81686611                        0  SP.POP.TOTL   
6              FRA  2020  67601110                        0  SP.POP.TOTL   
7              FRA  2019  67382061                        0  SP.POP.TOTL   
8              FRA  2018  67158348                        0  SP.POP.TOTL   
9              FRA  2017  66918020                        0  SP.POP.TOTL   
10             FRA  2016  66724104                        0  SP.POP.TOTL   
11             FRA  2015  66548272                        0  SP.POP.TOTL   

      indicator.value country.id country.value  
0   Population, total         DE       Germany  
1   Population, total         DE       Germany  
2   Population, total         DE       Germany  
3   Population, total         DE       Germany  
4   Population, total         DE       Germany  
5   Population, total         DE       Germany  
6   Population, total         FR        France  
7   Population, total         FR        France  
8   Population, total         FR        France  
9   Population, total         FR        France  
10  Population, total         FR        France  
11  Population, total         FR        France

### a)

#### The total population (SP.POP.TOTL) of Germany (DE) and France (FR) between 2015 and 2020.


In [141]:
df_POP_DE_FR = call_api(
    indicators="SP.POP.TOTL",
    countries=["DE","FR"],
    years="2015:2020"
)

url:  http://api.worldbank.org/v2/country/DE;FR/indicator/SP.POP.TOTL?date=2015:2020&format=json&per_page=10000&source=2


In [142]:
df_POP_DE_FR

countryiso3code  date     value unit obs_status  decimal indicator.id  \
0              DEU  2020  83160871                        0  SP.POP.TOTL   
1              DEU  2019  83092962                        0  SP.POP.TOTL   
2              DEU  2018  82905782                        0  SP.POP.TOTL   
3              DEU  2017  82657002                        0  SP.POP.TOTL   
4              DEU  2016  82348669                        0  SP.POP.TOTL   
5              DEU  2015  81686611                        0  SP.POP.TOTL   
6              FRA  2020  67601110                        0  SP.POP.TOTL   
7              FRA  2019  67382061                        0  SP.POP.TOTL   
8              FRA  2018  67158348                        0  SP.POP.TOTL   
9              FRA  2017  66918020                        0  SP.POP.TOTL   
10             FRA  2016  66724104                        0  SP.POP.TOTL   
11             FRA  2015  66548272                        0  SP.POP.TOTL   

      indicator.value country.id country.value  
0   Population, total         DE       Germany  
1   Population, total         DE       Germany  
2   Population, total         DE       Germany  
3   Population, total         DE       Germany  
4   Population, total         DE       Germany  
5   Population, total         DE       Germany  
6   Population, total         FR        France  
7   Population, total         FR        France  
8   Population, total         FR        France  
9   Population, total         FR        France  
10  Population, total         FR        France  
11  Population, total         FR        France

### b)

#### The total population (SP.POP.TOTL), GDP in current US$ (NY.GDP.MKTP.CD), and life expectancy in years at birth (SP.DYN.LE00.IN) of all countries (all) in 2012. Print the shape of the resulting DataFrame and display its first 10 rows


In [143]:
df_gdp_pop_2012 = call_api(
    indicators=["SP.POP.TOTL","NY.GDP.MKTP.CD","SP.DYN.LE00.IN"],
    countries='all',
    years="2012"
)

df_gdp_pop_2012

url:  http://api.worldbank.org/v2/country/all/indicator/SP.POP.TOTL;NY.GDP.MKTP.CD;SP.DYN.LE00.IN?date=2012&format=json&per_page=10000&source=2


countryiso3code  date         value unit obs_status  decimal  \
0               AFE  2012  5.596100e+08                        0   
1               AFW  2012  3.853603e+08                        0   
2               ARB  2012  3.810302e+08                        0   
3               CSS  2012  2.893251e+06                        0   
4               CEB  2012  1.039281e+08                        0   
..              ...   ...           ...  ...        ...      ...   
793             VIR  2012  7.851707e+01                        0   
794             PSE  2012  7.372200e+01                        0   
795             YEM  2012  6.756900e+01                        0   
796             ZMB  2012  5.870400e+01                        0   
797             ZWE  2012  5.538600e+01                        0   

       indicator.id                          indicator.value country.id  \
0       SP.POP.TOTL                        Population, total         ZH   
1       SP.POP.TOTL                        Population, total         ZI   
2       SP.POP.TOTL                        Population, total         1A   
3       SP.POP.TOTL                        Population, total         S3   
4       SP.POP.TOTL                        Population, total         B8   
..              ...                                      ...        ...   
793  SP.DYN.LE00.IN  Life expectancy at birth, total (years)         VI   
794  SP.DYN.LE00.IN  Life expectancy at birth, total (years)         PS   
795  SP.DYN.LE00.IN  Life expectancy at birth, total (years)         YE   
796  SP.DYN.LE00.IN  Life expectancy at birth, total (years)         ZM   
797  SP.DYN.LE00.IN  Life expectancy at birth, total (years)         ZW   

                      country.value scale  
0       Africa Eastern and Southern   NaN  
1        Africa Western and Central   NaN  
2                        Arab World   NaN  
3            Caribbean small states   NaN  
4    Central Europe and the Baltics   NaN  
..                              ...   ...  
793           Virgin Islands (U.S.)   NaN  
794              West Bank and Gaza   NaN  
795                     Yemen, Rep.   NaN  
796                          Zambia   NaN  
797                        Zimbabwe   NaN  

[798 rows x 11 columns]

In [144]:
df_gdp_pop_2012.to_csv("../data/gdp_pop_2012.csv", index=False)

### c)

#### State how many API calls your function makes for a) and b) respectively.


In [145]:
call_count

3

**Answer: for each call call_count variable counts number of calls in my case function is created fully adapted to type and number of parameters in url it's checking health of response additionally and provides solution based on function params**

- For this solution I researched API parameters such as sources which is set to two for fetching multiple indicator
- for seeing a lot of data I set per_page=10000 for fetching high amount of data


## Exercise 2

**The file medal_table_2012.csv contains information about the number of medals won by each country at the Olympic Games 2012. (It probably looks similar to the medal table that you calculated in the first part of the project. Small differences are possible, but the overall structure should be the same.)**

a) Preprocess both the medal table data and the Worldbank data retrieved in exercise 1 b) and
combine the two datasets suitably into one tidy dataset. The final dataset should be such that it
allows you to answer the following exercises. Explain your actions and decisions in a few sentences.

b) Create an alternative medal table for the 2012 Olympic Games by calculating the number of Gold,
Silver, and Bronze medals won per 10 million inhabitants. Display the 10 most successful countries
according to this alternative medal table.

Note: If there are missing values in the Worldbank data set (e.g. if no population data is available for
Germany), then you do NOT need to impute these values.


In [146]:
MEDAL_PATH = "../data/medal_table_2012.csv"

In [147]:
medal_table = pd.read_csv(MEDAL_PATH)


In [148]:
medal_table.head()

rank country_code                     country  gold  silver  bronze
0     1          USA               United States    46      28      30
1     2          CHN  People's Republic of China    38      31      22
2     3          GBR               Great Britain    29      17      19
3     4          RUS          Russian Federation    20      20      27
4     5          KOR           Republic of Korea    13       9       8

In [149]:
len(medal_table.country_code.unique())

85

In [150]:
medal_table.country_code.unique()

array(['USA', 'CHN', 'GBR', 'RUS', 'KOR', 'GER', 'FRA', 'AUS', 'ITA',
       'HUN', 'JPN', 'IRI', 'NZL', 'UKR', 'CUB', 'ESP', 'JAM', 'CZE',
       'RSA', 'PRK', 'BRA', 'POL', 'ETH', 'KAZ', 'CRO', 'CAN', 'BLR',
       'KEN', 'DEN', 'ROU', 'AZE', 'SUI', 'NOR', 'LTU', 'TUN', 'SWE',
       'COL', 'MEX', 'GEO', 'IRL', 'ARG', 'SRB', 'SLO', 'TTO', 'TUR',
       'DOM', 'TPE', 'LAT', 'ALG', 'BRN', 'GRN', 'BAH', 'UGA', 'VEN',
       'EGY', 'IND', 'MGL', 'THA', 'BUL', 'FIN', 'INA', 'SVK', 'BEL',
       'ARM', 'EST', 'MAS', 'PUR', 'BOT', 'CYP', 'GAB', 'GUA', 'MNE',
       'POR', 'UZB', 'GRE', 'QAT', 'SGP', 'AFG', 'CMR', 'HKG', 'KSA',
       'KUW', 'MAR', 'TJK', 'VIE'], dtype=object)

In [151]:
world_bank_table = pd.read_csv("../data/gdp_pop_2012.csv")

In [152]:
world_bank_table.head()

countryiso3code  date        value  unit  obs_status  decimal indicator.id  \
0             AFE  2012  559609961.0   NaN         NaN        0  SP.POP.TOTL   
1             AFW  2012  385360349.0   NaN         NaN        0  SP.POP.TOTL   
2             ARB  2012  381030177.0   NaN         NaN        0  SP.POP.TOTL   
3             CSS  2012    2893251.0   NaN         NaN        0  SP.POP.TOTL   
4             CEB  2012  103928128.0   NaN         NaN        0  SP.POP.TOTL   

     indicator.value country.id                   country.value  scale  
0  Population, total         ZH     Africa Eastern and Southern    NaN  
1  Population, total         ZI      Africa Western and Central    NaN  
2  Population, total         1A                      Arab World    NaN  
3  Population, total         S3          Caribbean small states    NaN  
4  Population, total         B8  Central Europe and the Baltics    NaN

In [153]:
world_bank_table['countryiso3code'].unique()

array(['AFE', 'AFW', 'ARB', 'CSS', 'CEB', 'EAR', 'EAS', 'EAP', 'TEA',
       'EMU', 'ECS', 'ECA', 'TEC', 'EUU', 'FCS', 'HPC', nan, 'IBD', 'IBT',
       'IDB', 'IDX', 'IDA', 'LTE', 'LCN', 'LAC', 'TLA', 'LDC', 'LMY',
       'MEA', 'MNA', 'TMN', 'MIC', 'NAC', 'OED', 'OSS', 'PSS', 'PST',
       'PRE', 'SST', 'SAS', 'TSA', 'SSF', 'SSA', 'TSS', 'WLD', 'AFG',
       'ALB', 'DZA', 'ASM', 'AND', 'AGO', 'ATG', 'ARG', 'ARM', 'ABW',
       'AUS', 'AUT', 'AZE', 'BHS', 'BHR', 'BGD', 'BRB', 'BLR', 'BEL',
       'BLZ', 'BEN', 'BMU', 'BTN', 'BOL', 'BIH', 'BWA', 'BRA', 'VGB',
       'BRN', 'BGR', 'BFA', 'BDI', 'CPV', 'KHM', 'CMR', 'CAN', 'CYM',
       'CAF', 'TCD', 'CHI', 'CHL', 'CHN', 'COL', 'COM', 'COD', 'COG',
       'CRI', 'CIV', 'HRV', 'CUB', 'CUW', 'CYP', 'CZE', 'DNK', 'DJI',
       'DMA', 'DOM', 'ECU', 'EGY', 'SLV', 'GNQ', 'ERI', 'EST', 'SWZ',
       'ETH', 'FRO', 'FJI', 'FIN', 'FRA', 'PYF', 'GAB', 'GMB', 'GEO',
       'DEU', 'GHA', 'GIB', 'GRC', 'GRL', 'GRD', 'GUM', 'GTM', 'GIN',
       'GNB', '

In [154]:
country_code_mapping = {
    'USA': 'US', # United States
    'GBR': 'GB', # Great Britain
    'CHN': 'CN', # China
    'RUS': 'RU', # Russia
    'KOR': 'KR', # Republic of Korea
    'GER': 'DE', # Germany
    'FRA': 'FR', # France
    'AUS': 'AU', # Australia
    'ITA': 'IT', # Italy
    'JPN': 'JP', # Japan
    'IRI': 'IR', # Iran
    'NZL': 'NZ', # New Zealand
    'UKR': 'UA', # Ukraine
    'CUB': 'CU', # Cuba
    'ESP': 'ES', # Spain
    'JAM': 'JM', # Jamaica
    'CZE': 'CZ', # Czech Republic
    'RSA': 'ZA', # South Africa
    'PRK': 'KP', # North Korea
    'BRA': 'BR', # Brazil
    'POL': 'PL', # Poland
    'ETH': 'ET', # Ethiopia
    'KAZ': 'KZ', # Kazakhstan
    'CRO': 'HR', # Croatia
    'CAN': 'CA', # Canada
    'BLR': 'BY', # Belarus
    'KEN': 'KE', # Kenya
    'DEN': 'DK', # Denmark
    'ROU': 'RO', # Romania
    'AZE': 'AZ', # Azerbaijan
    'SUI': 'CH', # Switzerland
    'NOR': 'NO', # Norway
    'LTU': 'LT', # Lithuania
    'TUN': 'TN', # Tunisia
    'SWE': 'SE', # Sweden
    'COL': 'CO', # Colombia
    'MEX': 'MX', # Mexico
    'GEO': 'GE', # Georgia
    'IRL': 'IE', # Ireland
    'ARG': 'AR', # Argentina
    'SRB': 'RS', # Serbia
    'SLO': 'SI', # Slovenia
    'TTO': 'TT', # Trinidad and Tobago
    'TUR': 'TR', # Turkey
    'DOM': 'DO', # Dominican Republic
    'TPE': 'TW', # Chinese Taipei
    'LAT': 'LV', # Latvia
    'ALG': 'DZ', # Algeria
    'BRN': 'BH', # Bahrain
    'GRN': 'GD', # Grenada
    'BAH': 'BS', # Bahamas
    'UGA': 'UG', # Uganda
    'VEN': 'VE', # Venezuela
    'EGY': 'EG', # Egypt
    'IND': 'IN', # India
    'MGL': 'MN', # Mongolia
    'THA': 'TH', # Thailand
    'BUL': 'BG', # Bulgaria
    'FIN': 'FI', # Finland
    'INA': 'ID', # Indonesia
    'SVK': 'SK', # Slovakia
    'BEL': 'BE', # Belgium
    'ARM': 'AM', # Armenia
    'EST': 'EE', # Estonia
    'MAS': 'MY', # Malaysia
    'PUR': 'PR', # Puerto Rico
    'BOT': 'BW', # Botswana
    'CYP': 'CY', # Cyprus
    'GAB': 'GA', # Gabon
    'GUA': 'GT', # Guatemala
    'MNE': 'ME', # Montenegro
    'POR': 'PT', # Portugal
    'UZB': 'UZ', # Uzbekistan
    'GRE': 'GR', # Greece
    'QAT': 'QA', # Qatar
    'SGP': 'SG', # Singapore
    'AFG': 'AF', # Afghanistan
    'CMR': 'CM', # Cameroon
    'HKG': 'HK', # Hong Kong
    'KSA': 'SA', # Saudi Arabia
    'KUW': 'KW', # Kuwait
    'MAR': 'MA', # Morocco
    'TJK': 'TJ', # Tajikistan
    'VIE': 'VN'  # Vietnam
}

In [155]:
medal_table['country_id'] = medal_table['country_code'].map(country_code_mapping)

# Display to check the mapping
medal_table[['country_code', 'country_id', 'country']].head(10)

country_code country_id                     country
0          USA         US               United States
1          CHN         CN  People's Republic of China
2          GBR         GB               Great Britain
3          RUS         RU          Russian Federation
4          KOR         KR           Republic of Korea
5          GER         DE                     Germany
6          FRA         FR                      France
7          AUS         AU                   Australia
8          ITA         IT                       Italy
9          HUN        NaN                     Hungary

In [156]:
world_bank_table.head()

countryiso3code  date        value  unit  obs_status  decimal indicator.id  \
0             AFE  2012  559609961.0   NaN         NaN        0  SP.POP.TOTL   
1             AFW  2012  385360349.0   NaN         NaN        0  SP.POP.TOTL   
2             ARB  2012  381030177.0   NaN         NaN        0  SP.POP.TOTL   
3             CSS  2012    2893251.0   NaN         NaN        0  SP.POP.TOTL   
4             CEB  2012  103928128.0   NaN         NaN        0  SP.POP.TOTL   

     indicator.value country.id                   country.value  scale  
0  Population, total         ZH     Africa Eastern and Southern    NaN  
1  Population, total         ZI      Africa Western and Central    NaN  
2  Population, total         1A                      Arab World    NaN  
3  Population, total         S3          Caribbean small states    NaN  
4  Population, total         B8  Central Europe and the Baltics    NaN

In [157]:
combined_df = pd.merge(
    medal_table, 
    world_bank_table, 
    left_on='country_id', 
    right_on='country.id', 
    how='inner'
)

# Check the merged result
combined_df.head()

rank country_code                     country  gold  silver  bronze  \
0     1          USA               United States    46      28      30   
1     1          USA               United States    46      28      30   
2     1          USA               United States    46      28      30   
3     2          CHN  People's Republic of China    38      31      22   
4     2          CHN  People's Republic of China    38      31      22   

  country_id countryiso3code  date         value  unit  obs_status  decimal  \
0         US             USA  2012  3.138777e+08   NaN         NaN        0   
1         US             USA  2012  1.625397e+13   NaN         NaN        0   
2         US             USA  2012  7.874146e+01   NaN         NaN        0   
3         CN             CHN  2012  1.354190e+09   NaN         NaN        0   
4         CN             CHN  2012  8.532185e+12   NaN         NaN        0   

     indicator.id                          indicator.value country.id  \
0     SP.POP.TOTL                        Population, total         US   
1  NY.GDP.MKTP.CD                        GDP (current US$)         US   
2  SP.DYN.LE00.IN  Life expectancy at birth, total (years)         US   
3     SP.POP.TOTL                        Population, total         CN   
4  NY.GDP.MKTP.CD                        GDP (current US$)         CN   

   country.value  scale  
0  United States    NaN  
1  United States    NaN  
2  United States    NaN  
3          China    NaN  
4          China    NaN

**Answer: Preprocessed data according medal table and data from API, for further analysis.**

**My approuch**

- There was problem with combination of tables country code didi not match with each other but I used dictionalry with matching conversion for IDs which help to create suitable format for IDs
- merging for combaning all together


**b) Create an alternative medal table for the 2012 Olympic Games by calculating the number of Gold, Silver, and Bronze medals won per 10 million inhabitants. Display the 10 most successful countries according to this alternative medal table.**


In [158]:
alternative_table = combined_df.copy()

In [159]:
alternative_table['total_medals'] = alternative_table['gold'] + alternative_table['silver'] + alternative_table['bronze']

In [160]:
alternative_table['gold_per_10m'] = alternative_table['gold'] * 10000000 / alternative_table['value']
alternative_table['silver_per_10m'] = alternative_table['silver'] * 10000000 / alternative_table['value']
alternative_table['bronze_per_10m'] = alternative_table['bronze'] * 10000000 / alternative_table['value']
alternative_table['total_per_10m'] = alternative_table['total_medals'] * 10000000 / alternative_table['value']


In [161]:
alt_medal_table = alternative_table.sort_values(
    by=['gold_per_10m', 'silver_per_10m', 'bronze_per_10m'], 
    ascending=False
)

In [162]:
alternative_table.head()

rank country_code                     country  gold  silver  bronze  \
0     1          USA               United States    46      28      30   
1     1          USA               United States    46      28      30   
2     1          USA               United States    46      28      30   
3     2          CHN  People's Republic of China    38      31      22   
4     2          CHN  People's Republic of China    38      31      22   

  country_id countryiso3code  date         value  ...    indicator.id  \
0         US             USA  2012  3.138777e+08  ...     SP.POP.TOTL   
1         US             USA  2012  1.625397e+13  ...  NY.GDP.MKTP.CD   
2         US             USA  2012  7.874146e+01  ...  SP.DYN.LE00.IN   
3         CN             CHN  2012  1.354190e+09  ...     SP.POP.TOTL   
4         CN             CHN  2012  8.532185e+12  ...  NY.GDP.MKTP.CD   

                           indicator.value  country.id  country.value scale  \
0                        Population, total          US  United States   NaN   
1                        GDP (current US$)          US  United States   NaN   
2  Life expectancy at birth, total (years)          US  United States   NaN   
3                        Population, total          CN          China   NaN   
4                        GDP (current US$)          CN          China   NaN   

  total_medals  gold_per_10m  silver_per_10m  bronze_per_10m  total_per_10m  
0          104  1.465539e+00    8.920673e-01    9.557864e-01   3.313393e+00  
1          104  2.830078e-05    1.722656e-05    1.845703e-05   6.398437e-05  
2          104  5.841903e+06    3.555941e+06    3.809937e+06   1.320778e+07  
3           91  2.806105e-01    2.289191e-01    1.624587e-01   6.719884e-01  
4           91  4.453724e-05    3.633301e-05    2.578472e-05   1.066550e-04  

[5 rows x 23 columns]

In [163]:
alternative_table.sort_values(by='total_per_10m', ascending=False, inplace=True)

In [164]:
alternative_table.head(10).sort_values(by='total_per_10m', ascending=False)

rank country_code                     country  gold  silver  bronze  \
2      1          USA               United States    46      28      30   
5      2          CHN  People's Republic of China    38      31      22   
11     4          RUS          Russian Federation    20      20      27   
8      3          GBR               Great Britain    29      17      19   
17     6          GER                     Germany    11      20      13   
32    11          JPN                       Japan     7      14      17   
20     7          FRA                      France    11      11      13   
23     8          AUS                   Australia     8      15      12   
14     5          KOR           Republic of Korea    13       9       8   
26     9          ITA                       Italy     8       9      11   

   country_id countryiso3code  date      value  ...    indicator.id  \
2          US             USA  2012  78.741463  ...  SP.DYN.LE00.IN   
5          CN             CHN  2012  76.200000  ...  SP.DYN.LE00.IN   
11         RU             RUS  2012  70.072195  ...  SP.DYN.LE00.IN   
8          GB             GBR  2012  80.904878  ...  SP.DYN.LE00.IN   
17         DE             DEU  2012  80.539024  ...  SP.DYN.LE00.IN   
32         JP             JPN  2012  83.096098  ...  SP.DYN.LE00.IN   
20         FR             FRA  2012  81.968293  ...  SP.DYN.LE00.IN   
23         AU             AUS  2012  82.046341  ...  SP.DYN.LE00.IN   
14         KR             KOR  2012  80.819512  ...  SP.DYN.LE00.IN   
26         IT             ITA  2012  82.239024  ...  SP.DYN.LE00.IN   

                            indicator.value  country.id       country.value  \
2   Life expectancy at birth, total (years)          US       United States   
5   Life expectancy at birth, total (years)          CN               China   
11  Life expectancy at birth, total (years)          RU  Russian Federation   
8   Life expectancy at birth, total (years)          GB      United Kingdom   
17  Life expectancy at birth, total (years)          DE             Germany   
32  Life expectancy at birth, total (years)          JP               Japan   
20  Life expectancy at birth, total (years)          FR              France   
23  Life expectancy at birth, total (years)          AU           Australia   
14  Life expectancy at birth, total (years)          KR         Korea, Rep.   
26  Life expectancy at birth, total (years)          IT               Italy   

   scale total_medals  gold_per_10m  silver_per_10m  bronze_per_10m  \
2    NaN          104  5.841903e+06    3.555941e+06    3.809937e+06   
5    NaN           91  4.986877e+06    4.068241e+06    2.887139e+06   
11   NaN           67  2.854199e+06    2.854199e+06    3.853169e+06   
8    NaN           65  3.584456e+06    2.101233e+06    2.348437e+06   
17   NaN           44  1.365798e+06    2.483268e+06    1.614124e+06   
32   NaN           38  8.423982e+05    1.684796e+06    2.045824e+06   
20   NaN           35  1.341982e+06    1.341982e+06    1.585979e+06   
23   NaN           35  9.750587e+05    1.828235e+06    1.462588e+06   
14   NaN           30  1.608522e+06    1.113592e+06    9.898600e+05   
26   NaN           28  9.727742e+05    1.094371e+06    1.337565e+06   

    total_per_10m  
2    1.320778e+07  
5    1.194226e+07  
11   9.561567e+06  
8    8.034126e+06  
17   5.463190e+06  
32   4.573019e+06  
20   4.269944e+06  
23   4.265882e+06  
14   3.711975e+06  
26   3.404710e+06  

[10 rows x 23 columns]

**Answer: Here Are presented list of most succesful countires by alterantive medal table based on amount of medlas per 10 Million inhabitants**

**My approuch**

- create alt table with total medals column
- compute alt ratio
- sort and print them
  `


## Exercise 3

**Carry out a simple supervised machine learning experiment, in which you train a model to predict the number of Gold medals a country wins at the Olympic Games 2012 based on demographic and economic features. Note: Since machine learning is not a focus topic of this course, you do not need to optimize the model. Just demonstrate that you are able to apply the steps we discussed in the course and correctly interpret the results.**

a) Train and evaluate a linear regression model: 1. Split your data into a training and a test set. 2. Train a linear regression model using population, life expectancy and the GDP per capita of a
country as features. 3. Evaluate the model using the root mean squared error as the performance
metric.

b) Discuss your results: How do you judge the performance? What are possible reasons for this
performance? How could the model be improved?

c) Due to an unfortunate “data error”, the country Netherlands was not included in the Olympic
Games data and is therefore not present in medal table. Use your trained machine learning model
to predict the number of Gold medals the Netherlands has won in 2012, just based on their demo-
graphic and economic characteristics.


a) Train and evaluate a linear regression model: 1. Split your data into a training and a test set. 2. Train a linear regression model using population, life expectancy and the GDP per capita of a
country as features. 3. Evaluate the model using the root mean squared error as the performance
metric.


In [165]:
from sklearn.metrics import root_mean_squared_error
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler

In [166]:
alternative_table.columns

Index(['rank', 'country_code', 'country', 'gold', 'silver', 'bronze',
       'country_id', 'countryiso3code', 'date', 'value', 'unit', 'obs_status',
       'decimal', 'indicator.id', 'indicator.value', 'country.id',
       'country.value', 'scale', 'total_medals', 'gold_per_10m',
       'silver_per_10m', 'bronze_per_10m', 'total_per_10m'],
      dtype='object')

In [167]:
alternative_table.describe()

rank        gold      silver      bronze    date         value  \
count  252.000000  252.000000  252.000000  252.000000   252.0  2.510000e+02   
mean    42.952381    3.511905    3.511905    4.107143  2012.0  2.769936e+11   
std     24.726506    7.522864    5.926833    5.832355     0.0  1.296281e+12   
min      1.000000    0.000000    0.000000    0.000000  2012.0  5.773300e+01   
25%     21.750000    0.000000    0.000000    1.000000  2012.0  8.037927e+01   
50%     42.500000    1.000000    1.000000    2.000000  2012.0  1.110693e+07   
75%     64.250000    3.000000    4.000000    4.250000  2012.0  4.301005e+10   
max     85.000000   46.000000   31.000000   30.000000  2012.0  1.625397e+13   

       unit  obs_status  decimal  scale  total_medals  gold_per_10m  \
count   0.0         0.0    252.0    0.0    252.000000  2.510000e+02   
mean    NaN         NaN      0.0    NaN     11.130952  1.542568e+05   
std     NaN         NaN      0.0    NaN     18.582719  6.019591e+05   
min     NaN         NaN      0.0    NaN      1.000000  0.000000e+00   
25%     NaN         NaN      0.0    NaN      2.000000  0.000000e+00   
50%     NaN         NaN      0.0    NaN      4.000000  2.217134e-04   
75%     NaN         NaN      0.0    NaN     11.000000  4.888361e+00   
max     NaN         NaN      0.0    NaN    104.000000  5.841903e+06   

       silver_per_10m  bronze_per_10m  total_per_10m  
count    2.510000e+02    2.510000e+02   2.510000e+02  
mean     1.533517e+05    1.812508e+05   4.888594e+05  
std      4.909935e+05    5.091693e+05   1.550564e+06  
min      0.000000e+00    0.000000e+00   1.347982e-05  
25%      0.000000e+00    4.938009e-05   8.635500e-04  
50%      3.067002e-03    3.821336e-01   4.672948e+00  
75%      1.813041e+01    1.318912e+05   2.459754e+05  
max      4.068241e+06    3.853169e+06   1.320778e+07

In [177]:
alternative_table.head()

rank country_code                     country  gold  silver  bronze  \
2      1          USA               United States    46      28      30   
5      2          CHN  People's Republic of China    38      31      22   
11     4          RUS          Russian Federation    20      20      27   
8      3          GBR               Great Britain    29      17      19   
17     6          GER                     Germany    11      20      13   

   country_id countryiso3code  date      value  ...    indicator.id  \
2          US             USA  2012  78.741463  ...  SP.DYN.LE00.IN   
5          CN             CHN  2012  76.200000  ...  SP.DYN.LE00.IN   
11         RU             RUS  2012  70.072195  ...  SP.DYN.LE00.IN   
8          GB             GBR  2012  80.904878  ...  SP.DYN.LE00.IN   
17         DE             DEU  2012  80.539024  ...  SP.DYN.LE00.IN   

                            indicator.value  country.id       country.value  \
2   Life expectancy at birth, total (years)          US       United States   
5   Life expectancy at birth, total (years)          CN               China   
11  Life expectancy at birth, total (years)          RU  Russian Federation   
8   Life expectancy at birth, total (years)          GB      United Kingdom   
17  Life expectancy at birth, total (years)          DE             Germany   

   scale total_medals  gold_per_10m  silver_per_10m  bronze_per_10m  \
2    NaN          104  5.841903e+06    3.555941e+06    3.809937e+06   
5    NaN           91  4.986877e+06    4.068241e+06    2.887139e+06   
11   NaN           67  2.854199e+06    2.854199e+06    3.853169e+06   
8    NaN           65  3.584456e+06    2.101233e+06    2.348437e+06   
17   NaN           44  1.365798e+06    2.483268e+06    1.614124e+06   

    total_per_10m  
2    1.320778e+07  
5    1.194226e+07  
11   9.561567e+06  
8    8.034126e+06  
17   5.463190e+06  

[5 rows x 23 columns]

In [179]:
features = alternative_table[['value','decimal']]
target= alternative_table['total_medals']
features.shape

(252, 2)

In [ ]:
features.isna().sum()
features = features.fillna(0) 


In [190]:
features.shape

(252, 2)

In [181]:
target.shape

(252,)

In [191]:
X_train, X_test, y_train, y_test = train_test_split(features, target, test_size=0.2, random_state=42)
scaler = StandardScaler()


In [192]:
X_train_scaled = scaler.fit_transform(X_train)

In [193]:
model = LinearRegression()
model.fit(X_train_scaled, y_train)
X_test_scaled = scaler.transform(X_test)
y_pred = model.predict(X_test_scaled)
rmse = np.sqrt(np.mean((y_test - y_pred) ** 2))

In [196]:
print("RMSE:", rmse)
print("Model Coefficients:", model.coef_)
print("Model Intercept:", model.intercept_)

RMSE: 12.471870135007567
Model Coefficients: [4.98912832 0.        ]
Model Intercept: 10.74129353233831


**Answer: Linear MOdel is trained on miserable amount of columns and inconsistente data from API**

**My approuch**

- event through feature dataset includ eonly population data which was precise gdp value column was total 0-s and life excpectance which was returned for me as a total NaN values data from API that is why I didint not included it in my features because handling of its data is absolutely different topic
- classic approuch preprocessing values in comfortable numeric format scaling, feeding to model and inferencing using RMSE


b) Discuss your results: How do you judge the performance? What are possible reasons for this
performance? How could the model be improved?


**Answer: Considering logic of whole project everything is well defined including denpendance of data, target of analysis and objective**

**My approuch**

- I think model could be improved firstly by adding valuable dataset which must include more columns more rows and valid points
- addingvariaty of models and fintuning approuches for exploring different approuches and results
- even through considering zero valued column for training, resutls are not bad 12.0 RMSE is accepted


c) Due to an unfortunate “data error”, the country Netherlands was not included in the Olympic
Games data and is therefore not present in medal table. Use your trained machine learning model
to predict the number of Gold medals the Netherlands has won in 2012, just based on their demo-
graphic and economic characteristics.


In [ ]:
netherlands = world_bank_table[world_bank_table['country.value'] == 'Netherlands']
netherlands_feature = netherlands[['value','decimal']]
netherlands_feature = netherlands_feature.fillna(0)


Netherlands Predicted Total Medals: 9.180667591196656


In [206]:
netherlands_feature_scaled = scaler.transform(netherlands_feature)
netherlands_pred = model.predict(netherlands_feature_scaled)


In [207]:
print("Netherlands Predicted Total Medals:", netherlands_pred[0])

Netherlands Predicted Total Medals: 9.180667591196656


**Answer: Case for Netherlands accepted and solved accordingly**

**My approuch**

- Firstly I wrangle data realted to Netherland form original api returned data of world bank data
- Then I filter to take features which are used in predicition
- scaled ofc and processed predicting which gave in result 9.18 which is realistic approximation and means that in 2012 Netherland could won 9 medals in case if me round answered value down
